### PR TITLE
Feature/fix and improviment tests

### PR DESCRIPTION
**Pull request by dexfs**
I changed AbstractResource, Users and Auth Classes. The relatives tests too 

There are some points I need to see with you.  
- How the better way to inform endpoint, current each endpoint is informed on the method. 
- What class is responsible to keep the AUTH TOKEN after a success authentication, current the abstractResource keeping it

### CODE REVIEW

File: composer.json [hunk 1]
```diff
@@ -1,20 +1,34 @@
 {
+  "name": "HookOps/hubstaff-php",
+  "type": "library",
+  "description": "A PHP toolkit for Hubstaff API",
+  "homepage": "https://github.com/HookOps/hubstaff-php",
+  "keywords": [
+    "php",
```
Comment (malukenho) [inline composer.json]: No need for `php` here, as `composer` is a package manager for php, It's redundant.

File: src/AbstractResource.php [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+
+    /**
+     * @var DecodeDataInterface
+     */
+    protected $decoder;
+
+    /**
+     * @var string
+     */
+    protected $appToken = 'string';
```
Comment (malukenho) [inline src/AbstractResource.php]: there's any other possibility for `appToken` type?

File: src/AbstractResource.php [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+    {
+        return $this->decoder->decode($this->client->send($method, $url, $headers, $parameters));
+    }
+
+    /**
+     * @return array
+     */
+    protected function getHeaders()
+    {
+        return [
```
Comment (malukenho) [inline src/AbstractResource.php]: IMHO this tokens shall be a value object

File: src/Activities.php [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+     * @param string $auth_token
+     * @param string $app_token
+     * @param string $starttime
+     * @param string $endtime
+     * @param int $offset
+     * @param array $options
+     * @param string $url
+     * @return array
+     */
+    public function getActivities($auth_token, $app_token, $starttime, $endtime, $offset, $options, $url)
```
Comment (malukenho) [inline src/Activities.php]: too much parameters... use sneak case please?

File: src/Activities.php [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+
+        $parameters['Auth-Token'] = 'header';
+        $parameters['App-token'] = 'header';
+        $parameters['start_time'] = '';
+        $parameters['stop_time'] = '';
+        $parameters['offset'] = '';
+
+        return $this->returnDecodedData($url, $fields, $parameters);
+    }
+}
```
Comment (malukenho) [inline src/Activities.php]: EOL EOF

File: src/Auth.php [hunk 6]
```diff
@@ -0,0 +1,20 @@
+<?php namespace Hubstaff;
+
+final class Auth extends AbstractResource
+{
+    public function auth($email, $password)
+    {
+        $parameters['email'] = $email;
```
Comment (malukenho) [inline src/Auth.php]: Where did `$parameters` come from? Are your development environment `display_errors` and `error_reporting(-1);` enabled?

File: src/Activities.php [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+     * @param array $options
+     * @param string $url
+     * @return array
+     */
+    public function getActivities($auth_token, $app_token, $starttime, $endtime, $offset, $options, $url)
+    {
+        $fields['Auth-Token'] = $auth_token;
+        $fields['App-token'] = $app_token;
+        $fields['start_time'] = $starttime;
+        $fields['stop_time'] = $endtime;
```
Comment (malukenho) [inline src/Activities.php]: What about we have fields and options as VO's as well?

File: src/Auth.php [hunk 8]
```diff
@@ -0,0 +1,20 @@
...
+        $parameters['password'] = $password;
+
+        return $this->returnDecodedData('POST', '/v1/auth', $parameters, $this->getHeaders());
+    }
+
+    protected function getHeaders()
```
Comment (malukenho) [inline src/Auth.php]: Think about Headers be other VO as well...

File: src/Auth.php [hunk 9]
```diff
@@ -0,0 +1,20 @@
...
+    {
+        return [
+            'App-Token'  => $this->appToken,
+        ];
+    }
+}
+
```
Comment (malukenho) [inline src/Auth.php]: Too much line in it

File: src/Client.php [hunk 10]
```diff
@@ -0,0 +1,314 @@
+<?php
+
+namespace Hubstaff;
```
Comment (malukenho) [inline src/Client.php]: I'm into use a separated line for namespaces, but I'm seeing you are inconsistent with that here. Can you considere using `phpcs`?

File: src/Client.php [hunk 11]
```diff
@@ -0,0 +1,314 @@
...
+
+final class Client
+{
+    /**
+     * @var null
+     */
+    private $app_token = null;
+    /**
```
Comment (malukenho) [inline src/Client.php]: space in between

File: src/Client.php [hunk 12]
```diff
@@ -0,0 +1,314 @@
+<?php
+
+namespace Hubstaff;
+
+final class Client
+{
+    /**
+     * @var null
```
Comment (malukenho) [inline src/Client.php]: `null`?

File: src/Client.php [hunk 13]
```diff
@@ -0,0 +1,314 @@
...
+     */
+    private $app_token = null;
+    /**
+     * @var null
```
Comment (malukenho) [inline src/Client.php]: seems wrong.

File: src/Client.php [hunk 14]
```diff
@@ -0,0 +1,314 @@
...
+     */
+    private $auth_token = null;
+
+
```
Comment (malukenho) [inline src/Client.php]: too much blank line

File: src/Client.php [hunk 15]
```diff
@@ -0,0 +1,314 @@
...
+    /**
+     * Client constructor.
```
Comment (malukenho) [inline src/Client.php]: remove this comment

File: src/Client.php [hunk 16]
```diff
@@ -0,0 +1,314 @@
...
+     * @param $app_token
+     */
+    public function __construct($app_token)
+    {
+        $this->app_token = $app_token;
+    }
+
+    /**
+     * @return null
```
Comment (malukenho) [inline src/Client.php]: `nil`?

File: src/Client.php [hunk 17]
```diff
@@ -0,0 +1,314 @@
...
+     */
+    public function getAuthToken()
+    {
+        return $this->auth_token;
+    }
+
+    /**
+     * @param $auth_token
```
Comment (malukenho) [inline src/Client.php]: `@param string $auth_token`

File: src/Client.php [hunk 18]
```diff
@@ -0,0 +1,314 @@
...
+    {
+        $this->auth_token = $auth_token;
+    }
+
+    /**
+     * @param $email
+     * @param $password
+     * @return mixed
+     */
+    public function auth($email, $password)
```
Comment (malukenho) [inline src/Client.php]: Humm... return seems to be wrong, can we thrown an exception instead?

File: src/Client.php [hunk 19]
```diff
@@ -0,0 +1,314 @@
...
+        $this->setAuthToken($auth_token["auth_token"]);
+    }
+
+    /**
+     * @param int $organization_memberships
+     * @param int $project_memberships
+     * @param int $offset
+     * @return mixed
+     */
+    public function users($organization_memberships = 0, $project_memberships = 0, $offset = 0)
```
Comment (malukenho) [inline src/Client.php]: `0` is a valid option?

File: src/Decoder/DecodeDataInterface.php [hunk 20]
```diff
@@ -0,0 +1,6 @@
+<?php namespace Hubstaff\Decoder;
+
+interface DecodeDataInterface
+{
+    public function decode($data);
+}
```
Comment (malukenho) [inline src/Decoder/DecodeDataInterface.php]: EOL EOF

File: tests/WeeklyTest.php [hunk 21]
```diff
@@ -0,0 +1,40 @@
...
+    public function testWeekly_my()
+    {
+        $expected = json_decode('{"organizations":[]}', true);
+        $this->stub->expects($this->any())
+            ->method('weeklyMy')
+            ->will($this->returnValue($expected));
+
+        $this->assertArrayHasKey('organizations', $this->stub->weeklyMy($this->options));
+    }
+}
```
Comment (malukenho) [inline tests/WeeklyTest.php]: EOL EOF

File: tests/WeeklyTest.php [hunk 22]
```diff
@@ -0,0 +1,40 @@
+<?php
+
+class WeeklyTest extends \PHPUnit_Framework_TestCase
+{
+    private $stub;
+    private $options = [];
+
+    public function __construct()
```
Comment (malukenho) [inline tests/WeeklyTest.php]: `__construct`?  You meant `__setUp` right?

File: tests/WeeklyTest.php [hunk 23]
```diff
@@ -0,0 +1,40 @@
...
+    {
+        parent::__construct();
```
Comment (malukenho) [inline tests/WeeklyTest.php]: kill this line

File: tests/WeeklyTest.php [hunk 24]
```diff
@@ -0,0 +1,40 @@
...
+        $this->options['users'] = '61188';
```
Comment (malukenho) [inline tests/WeeklyTest.php]: `random_int(0, PHP_INT_MAX)`

File: tests/WeeklyTest.php [hunk 25]
```diff
@@ -0,0 +1,40 @@
...
+        $this->options['projects'] = '112761';
```
Comment (malukenho) [inline tests/WeeklyTest.php]: `random_int(0, PHP_INT_MAX)`

File: tests/WeeklyTest.php [hunk 26]
```diff
@@ -0,0 +1,40 @@
...
+        $this->options['organizations'] = '27572';
```
Comment (malukenho) [inline tests/WeeklyTest.php]: `random_int(0, PHP_INT_MAX)`

File: tests/WeeklyTest.php [hunk 27]
```diff
@@ -0,0 +1,40 @@
...
+        $this->options['date'] = '2016-05-01';
```
Comment (malukenho) [inline tests/WeeklyTest.php]: What if it's not a valid date?

File: tests/WeeklyTest.php [hunk 28]
```diff
@@ -0,0 +1,40 @@
...
+
+        $this->stub = $this->getMockBuilder('Hubstaff\Client')
```
Comment (malukenho) [inline tests/WeeklyTest.php]: `::class`

File: tests/WeeklyTest.php [hunk 29]
```diff
@@ -0,0 +1,40 @@
...
+    }
+
+    public function test_weekly_team()
+    {
+        $expected = json_decode('{"organizations":[]}', true);
+        $this->stub->expects($this->any())
+            ->method('weeklyTeam')
+            ->will($this->returnValue($expected));
+
+        $this->assertArrayHasKey('organizations', $this->stub->weeklyTeam($this->options));
```
Comment (malukenho) [inline tests/WeeklyTest.php]: `self::`

File: tests/WeeklyTest.php [hunk 30]
```diff
@@ -0,0 +1,40 @@
+<?php
+
+class WeeklyTest extends \PHPUnit_Framework_TestCase
+{
+    private $stub;
```
Comment (malukenho) [inline tests/WeeklyTest.php]: `doc`?

File: tests/WeeklyTest.php [hunk 31]
```diff
@@ -0,0 +1,40 @@
+<?php
+
+class WeeklyTest extends \PHPUnit_Framework_TestCase
```
Comment (malukenho) [inline tests/WeeklyTest.php]: `final`

File: tests/WeeklyTest.php [hunk 32]
```diff
@@ -0,0 +1,40 @@
+<?php
+
```
Comment (malukenho) [inline tests/WeeklyTest.php]: `@covers`

File: tests/WeeklyTest.php (hunk 31 above)
Comment (malukenho) [inline tests/WeeklyTest.php]: import `PHPUnit_Framework_TestCase`

File: tests/WeeklyTest.php (hunk 32 above)
Comment (malukenho) [inline tests/WeeklyTest.php]: Where's the tests namespaces?

File: tests/WeeklyTest.php [hunk 33]
```diff
@@ -0,0 +1,40 @@
...
+class WeeklyTest extends \PHPUnit_Framework_TestCase
+{
+    private $stub;
+    private $options = [];
```
Comment (malukenho) [inline tests/WeeklyTest.php]: docs?

File: tests/WeeklyTest.php [hunk 34]
```diff
@@ -0,0 +1,40 @@
...
+        $this->options['projects'] = '112761';
+        $this->options['organizations'] = '27572';
+        $this->options['date'] = '2016-05-01';
+
+        $this->stub = $this->getMockBuilder('Hubstaff\Client')
+            ->disableOriginalConstructor()
+            ->getMock();
+    }
+
+    public function test_weekly_team()
```
Comment (malukenho) [inline tests/WeeklyTest.php]: `test_weekly_team` doesn't say too much. What's it testing?

File: tests/WeeklyTest.php [hunk 35]
```diff
@@ -0,0 +1,40 @@
...
+    {
+        $expected = json_decode('{"organizations":[]}', true);
+        $this->stub->expects($this->any())
```
Comment (malukenho) [inline tests/WeeklyTest.php]: `self::any()`

File: tests/WeeklyTest.php [hunk 36]
```diff
@@ -0,0 +1,40 @@
...
+    }
+
+    public function testWeekly_my()
+    {
+        $expected = json_decode('{"organizations":[]}', true);
+        $this->stub->expects($this->any())
+            ->method('weeklyMy')
+            ->will($this->returnValue($expected));
+
+        $this->assertArrayHasKey('organizations', $this->stub->weeklyMy($this->options));
```
Comment (malukenho) [inline tests/WeeklyTest.php]: `self::`

File: tests/WeeklyTest.php [hunk 37]
```diff
@@ -0,0 +1,40 @@
...
+            ->method('weeklyTeam')
+            ->will($this->returnValue($expected));
+
+        $this->assertArrayHasKey('organizations', $this->stub->weeklyTeam($this->options));
+    }
+
+    public function testWeekly_my()
+    {
+        $expected = json_decode('{"organizations":[]}', true);
+        $this->stub->expects($this->any())
```
Comment (malukenho) [inline tests/WeeklyTest.php]: `self::any()`

File: tests/WeeklyTest.php [hunk 38]
```diff
@@ -0,0 +1,40 @@
...
+            ->method('weeklyTeam')
+            ->will($this->returnValue($expected));
+
+        $this->assertArrayHasKey('organizations', $this->stub->weeklyTeam($this->options));
+    }
+
+    public function testWeekly_my()
```
Comment (malukenho) [inline tests/WeeklyTest.php]: be consistent. `test_weekly_my`. also this test name doesn't say too much.

File: tests/WeeklyTest.php [hunk 39]
```diff
@@ -0,0 +1,40 @@
...
+    {
+        $expected = json_decode('{"organizations":[]}', true);
```
Comment (malukenho) [inline tests/WeeklyTest.php]: why not use an array directly?

File: tests/WeeklyTest.php [hunk 40]
```diff
@@ -0,0 +1,40 @@
...
+        $this->options['date'] = '2016-05-01';
+
+        $this->stub = $this->getMockBuilder('Hubstaff\Client')
+            ->disableOriginalConstructor()
+            ->getMock();
+    }
+
+    public function test_weekly_team()
+    {
+        $expected = json_decode('{"organizations":[]}', true);
```
Comment (malukenho) [inline tests/WeeklyTest.php]: just use an array directly

File: tests/UsersTest.php [hunk 41]
```diff
@@ -0,0 +1,183 @@
...
+            ->will(self::returnValue($expected));
+
+        $user = new Users($this->client, $this->decoder);
+        $response = $user->findUserProjects($id, $offset);
+
+        self::assertEquals($expected, $response);
+        self::assertEquals(array_keys($expected), array_keys($response));
+    }
+}
+
```
Comment (malukenho) [inline tests/UsersTest.php]: too much space

File: tests/UsersTest.php [hunk 42]
```diff
@@ -0,0 +1,183 @@
...
+    public function it_can_find_user_projects()
+    {
+        $id         = random_int(1, 20);
+        $url        = sprintf('/v1/users/%d/projects', $id);
+        $method     = 'GET';
+        $offset     = random_int(1, 10);
+
+        $headers = [
+            'App-Token' => $this->appToken,
+            'Auth-Token'=> $this->authToken
```
Comment (malukenho) [inline tests/UsersTest.php]: `,`

File: tests/UsersTest.php [hunk 43]
```diff
@@ -0,0 +1,183 @@
...
+        ];
+
+        $parameters = [
+            'offset' => $offset
```
Comment (malukenho) [inline tests/UsersTest.php]: `,`

File: tests/UsersTest.php [hunk 44]
```diff
@@ -0,0 +1,183 @@
...
```
Comment (malukenho) [inline tests/UsersTest.php]: `,`

File: tests/UsersTest.php [hunk 45]
```diff
@@ -0,0 +1,183 @@
...
+    public function it_can_find_user_orgs()
+    {
+        $id         = random_int(1, 20);
+        $url        = sprintf('/v1/users/%d/organizations', $id);
+        $method     = 'GET';
+        $offset     = random_int(1, 10);
+
+        $headers = [
+            'App-Token' => $this->appToken,
+            'Auth-Token'=> $this->authToken
```
Comment (malukenho) [inline tests/UsersTest.php]: `,`

File: tests/UsersTest.php [hunk 46]
```diff
@@ -0,0 +1,183 @@
...
+
+namespace Hubstaff;
+
+use Hubstaff\Decoder\DecodeDataInterface;
+use Hubstaff\Helper\ClientInterface;
+
+/**
+ * @covers \Hubstaff\Users
+ */
+class UsersTest extends \PHPUnit_Framework_TestCase
```
Comment (malukenho) [inline tests/UsersTest.php]: `final`

Comment (malukenho) [inline tests/UsersTest.php]: Import phpunit namespace

File: tests/UsersTest.php [hunk 47]
```diff
@@ -0,0 +1,183 @@
+<?php
+
+namespace Hubstaff;
```
Comment (malukenho) [inline tests/UsersTest.php]: `HubstaffTest` nope? See `composer` config for `require-dev`

File: tests/ScreenshotsTest.php [hunk 48]
```diff
@@ -0,0 +1,34 @@
...
+        $starttime = '2016-03-14';
+        $stoptime = '2016-03-20';
+        $expected = json_decode('{"screenshots":[]}', true);
+        $this->stub->expects($this->any())
+            ->method('screenshots')
+            ->will($this->returnValue($expected));
+
+        $this->assertArrayHasKey('screenshots', $this->stub->screenshots($starttime, $stoptime, $this->options, 0));
+    }
+}
```
Comment (malukenho) [inline tests/ScreenshotsTest.php]: EOL EOF

File: tests/ScreenshotsTest.php [hunk 49]
```diff
@@ -0,0 +1,34 @@
...
+            ->getMock();
+
+    }
+
+    public function test_screenshots()
+    {
+        $starttime = '2016-03-14';
+        $stoptime = '2016-03-20';
+        $expected = json_decode('{"screenshots":[]}', true);
+        $this->stub->expects($this->any())
```
Comment (malukenho) [inline tests/ScreenshotsTest.php]: `self::any()`

File: tests/ScreenshotsTest.php [hunk 50]
```diff
@@ -0,0 +1,34 @@
...
+    {
+        parent::__construct();
+        $this->options['users'] = '61188';
+        $this->options['projects'] = '112761';
+        $this->options['organizations'] = '27572';
+
+        $this->stub = $this->getMockBuilder('Hubstaff\Client')
+            ->disableOriginalConstructor()
+            ->getMock();
+
```
Comment (malukenho) [inline tests/ScreenshotsTest.php]: kill this line

File: tests/ScreenshotsTest.php [hunk 51]
```diff
@@ -0,0 +1,34 @@
+<?php
+
+namespace Hubstaff;
```
Comment (malukenho) [inline tests/ScreenshotsTest.php]: wrong namespace

File: tests/ScreenshotsTest.php [hunk 52]
```diff
@@ -0,0 +1,34 @@
...
+
+class ScreenshotsTest extends \PHPUnit_Framework_TestCase
```
Comment (malukenho) [inline tests/ScreenshotsTest.php]: `@covers`

File: tests/ScreenshotsTest.php [hunk 53]
```diff
@@ -0,0 +1,34 @@
...
+{
+    private $stub;
+    private $options = [];
+
+    public function __construct()
```
Comment (malukenho) [inline tests/ScreenshotsTest.php]: `__construct`?

File: tests/ScreenshotsTest.php [hunk 54]
```diff
@@ -0,0 +1,34 @@
...
+    {
+        parent::__construct();
+        $this->options['users'] = '61188';
```
Comment (malukenho) [inline tests/ScreenshotsTest.php]: `random_int(0, PHP_INT_MAX)`

File: tests/ScreenshotsTest.php [hunk 55]
```diff
@@ -0,0 +1,34 @@
...
+        $this->options['projects'] = '112761';
```
Comment (malukenho) [inline tests/ScreenshotsTest.php]: `random_int(0, PHP_INT_MAX)`

File: tests/ScreenshotsTest.php [hunk 56]
```diff
@@ -0,0 +1,34 @@
...
+        $this->options['organizations'] = '27572';
```
Comment (malukenho) [inline tests/ScreenshotsTest.php]: `random_int(0, PHP_INT_MAX)`

File: tests/ScreenshotsTest.php [hunk 57]
```diff
@@ -0,0 +1,34 @@
...
+
+        $this->stub = $this->getMockBuilder('Hubstaff\Client')
```
Comment (malukenho) [inline tests/ScreenshotsTest.php]: `::class`

File: tests/ScreenshotsTest.php [hunk 58]
```diff
@@ -0,0 +1,34 @@
+<?php
+
+namespace Hubstaff;
+
+class ScreenshotsTest extends \PHPUnit_Framework_TestCase
+{
+    private $stub;
```
Comment (malukenho) [inline tests/ScreenshotsTest.php]: `doc`?

File: tests/ProjectsTest.php [hunk 59]
```diff
@@ -0,0 +1,126 @@
...
+            ->method('send')
+            ->with($fields, $parameters, $url, 0)
+            ->will(self::returnValue([]));
+
+        $this->decoder->expects(self::once())->method('decode');
+
+        $project = new \Hubstaff\Projects($this->client, $this->decoder);
+        $project->findProjectMembers($authToken, $appToken, $offset, $url);
+    }
+}
```
Comment (malukenho) [inline tests/ProjectsTest.php]: EOL EOF

File: tests/ProjectsTest.php [hunk 60]
```diff
@@ -0,0 +1,126 @@
+<?php
+
+use Hubstaff\Decoder\DecodeDataInterface;
+use Hubstaff\Helper\ClientInterface;
+
+/**
+ * @cover \Hubstaff\Projects
```
Comment (malukenho) [inline tests/ProjectsTest.php]: `@covers`

File: tests/ProjectsTest.php [hunk 61]
```diff
@@ -0,0 +1,126 @@
+<?php
+
```
Comment (malukenho) [inline tests/ProjectsTest.php]: Test namespace

File: tests/ProjectsTest.php [hunk 62]
```diff
@@ -0,0 +1,126 @@
...
+use Hubstaff\Decoder\DecodeDataInterface;
+use Hubstaff\Helper\ClientInterface;
+
+/**
+ * @cover \Hubstaff\Projects
+ */
+class ProjectsTest extends \PHPUnit_Framework_TestCase
```
Comment (malukenho) [inline tests/ProjectsTest.php]: `final`

Comment (malukenho) [inline tests/ProjectsTest.php]: Import `PHPUnit_Framework_TestCase`

File: tests/ProjectsTest.php [hunk 63]
```diff
@@ -0,0 +1,126 @@
...
+
+        $this->client->expects(self::once())
+            ->method('send')
+            ->with($fields, $parameters, $url, 0)
+            ->will(self::returnValue([]));
+
+        $this->decoder->expects(self::once())->method('decode');
+        $project = new \Hubstaff\Projects($this->client, $this->decoder);
+        $project->findProject($authToken, $appToken, $url);
+
```
Comment (malukenho) [inline tests/ProjectsTest.php]: kill this line

File: tests/OrganizationsTest.php [hunk 64]
```diff
@@ -0,0 +1,57 @@
...
+    {
+        $expected = json_decode('{ "users": [ { "id": 61188, "name": "Raymond Cudjoe", "last_activity": "2016-05-24T01:25:21Z", "email": "rkcudjoe@hookengine.com", "pay_rate": "No rate set" } ] }', true);
+        $this->stub->expects($this->any())
+            ->method('findOrgMembers')
+            ->will($this->returnValue($expected));
+
+        $this->assertArrayHasKey('users', $this->stub->findOrgMembers(27572));
+    }
+}
+
```
Comment (malukenho) [inline tests/OrganizationsTest.php]: remove line

File: tests/OrganizationsTest.php [hunk 65]
```diff
@@ -0,0 +1,57 @@
+<?php
+
+namespace Hubstaff;
+
+class OrganizationsTest extends \PHPUnit_Framework_TestCase
+{
+    private $stub;
+
+    public function __construct()
```
Comment (malukenho) [inline tests/OrganizationsTest.php]: kill `__construct` usage

File: tests/OrganizationsTest.php [hunk 66]
```diff
@@ -0,0 +1,57 @@
...
+    {
+        parent::__construct();
```
Comment (malukenho) [inline tests/OrganizationsTest.php]: nope

File: tests/OrganizationsTest.php [hunk 67]
```diff
@@ -0,0 +1,57 @@
...
+        $this->stub = $this->getMockBuilder('Hubstaff\Client')->disableOriginalConstructor()->getMock();
```
Comment (malukenho) [inline tests/OrganizationsTest.php]: `::class`

File: tests/OrganizationsTest.php [hunk 68]
```diff
@@ -0,0 +1,57 @@
...
+    }
+
+    public function test_organizations()
+    {
+        $expected = json_decode('{ "organizations": [ { "id": 27572, "name": "Hook Engine", "last_activity": "2016-05-24T01:25:21Z" } ] }', true);
+        $this->stub->expects($this->any())
```
Comment (malukenho) [inline tests/OrganizationsTest.php]: `self::any()`

File: tests/OrganizationsTest.php [hunk 69]
```diff
@@ -0,0 +1,57 @@
...
+    }
+
+    public function test_find_org_members()
+    {
+        $expected = json_decode('{ "users": [ { "id": 61188, "name": "Raymond Cudjoe", "last_activity": "2016-05-24T01:25:21Z", "email": "rkcudjoe@hookengine.com", "pay_rate": "No rate set" } ] }', true);
+        $this->stub->expects($this->any())
+            ->method('findOrgMembers')
+            ->will($this->returnValue($expected));
+
+        $this->assertArrayHasKey('users', $this->stub->findOrgMembers(27572));
```
Comment (malukenho) [inline tests/OrganizationsTest.php]: `random_int(0, PHP_INT_MAX)`

File: tests/NotesTest.php [hunk 70]
```diff
@@ -0,0 +1,108 @@
...
+        $this->client->expects(self::once())
+            ->method('send')
+            ->with($fields, $parameters, $url, 0)
+            ->will(self::returnValue([]));
+
+        $this->decoder->expects(self::once())->method('decode');
+        $notes = new Notes($this->client, $this->decoder);
+        $notes->findNote($authToken, $appToken, $url);
+    }
+}
```
Comment (malukenho) [inline tests/NotesTest.php]: EOL EOF

File: tests/NotesTest.php [hunk 71]
```diff
@@ -0,0 +1,108 @@
...
+
+namespace Hubstaff;
+
+use Hubstaff\Decoder\DecodeDataInterface;
+use Hubstaff\Helper\ClientInterface;
+
+/**
+ * @covers \Hubstaff\Notes
+ */
+class NotesTest extends \PHPUnit_Framework_TestCase
```
Comment (malukenho) [inline tests/NotesTest.php]: `final`

Comment (malukenho) [inline tests/NotesTest.php]: Import `PHPUnit_Framework_TestCase`

File: tests/NotesTest.php [hunk 72]
```diff
@@ -0,0 +1,108 @@
+<?php
+
+namespace Hubstaff;
```
Comment (malukenho) [inline tests/NotesTest.php]: Test namespace

File: tests/CustomTest.php [hunk 73]
```diff
@@ -0,0 +1,141 @@
...
+
+        $custom = new Custom($this->client, $this->decoder);
+
+        $actual = $custom->customReport(
+            'auth_token',
+            'app_token',
+            'start_date',
+            'end_date',
+            'options',
+            'url'
```
Comment (malukenho) [inline tests/CustomTest.php]: `,`

File: tests/AuthTest.php [hunk 74]
```diff
@@ -0,0 +1,141 @@
...
+        $this->decoder->expects(self::once())
+            ->method('decode')
+            ->willReturn($expected);
+
+        $auth = new Auth($this->client, $this->decoder);
+        $response = $auth->auth($this->email, $this->password);
+
+        $this->assertEquals($expected, $response);
+    }
+}
```
Comment (malukenho) [inline tests/AuthTest.php]: EOL EOF

File: tests/AuthTest.php [hunk 75]
```diff
@@ -0,0 +1,141 @@
...
+
+        $auth = new Auth($this->client, $this->decoder);
+        $auth->auth($this->email, $this->password);
+    }
+
+
+    /**
+     * @test
+     */
+
```
Comment (malukenho) [inline tests/AuthTest.php]: test is floating in the limbo

File: tests/AuthTest.php [hunk 76]
```diff
@@ -0,0 +1,141 @@
...
+            ->with($method, $url, $parameters, $headers)
+            ->will(self::returnValue([]));
+
+        $this->decoder->expects(self::once())->method('decode');
+
+        $auth = new Auth($this->client, $this->decoder);
+        $auth->auth($this->email, $this->password);
+    }
+
+
```
Comment (malukenho) [inline tests/AuthTest.php]: remove blank line

File: tests/AuthTest.php [hunk 77]
```diff
@@ -0,0 +1,141 @@
...
+    /**
+     * @var DecodeDataInterface|\PHPUnit_Framework_MockObject_MockObject
+     */
+    private $decoder;
+
+    private $appToken = 'string';
+
+    private $email;
+
+    private $password;
```
Comment (malukenho) [inline tests/AuthTest.php]: missing doc

File: tests/AuthTest.php [hunk 78]
```diff
@@ -0,0 +1,141 @@
...
+    private $client;
+
+    /**
+     * @var DecodeDataInterface|\PHPUnit_Framework_MockObject_MockObject
+     */
+    private $decoder;
+
+    private $appToken = 'string';
+
+    private $email;
```
Comment (malukenho) [inline tests/AuthTest.php]: missing doc

File: tests/AuthTest.php [hunk 79]
```diff
@@ -0,0 +1,141 @@
...
+     * @var ClientInterface|\PHPUnit_Framework_MockObject_MockObject
+     */
+    private $client;
+
+    /**
+     * @var DecodeDataInterface|\PHPUnit_Framework_MockObject_MockObject
+     */
+    private $decoder;
+
+    private $appToken = 'string';
```
Comment (malukenho) [inline tests/AuthTest.php]: missing doc

File: tests/AuthTest.php [hunk 80]
```diff
@@ -0,0 +1,141 @@
+<?php
+
+use Hubstaff\Auth;
+use Hubstaff\Decoder\DecodeDataInterface;
+use Hubstaff\Helper\ClientInterface;
+
+class AuthTest extends PHPUnit_Framework_TestCase
+{
+
```
Comment (malukenho) [inline tests/AuthTest.php]: remove blank line

File: tests/AuthTest.php [hunk 81]
```diff
@@ -0,0 +1,141 @@
+<?php
+
```
Comment (malukenho) [inline tests/AuthTest.php]: namespace?

File: tests/AuthTest.php [hunk 82]
```diff
@@ -0,0 +1,141 @@
...
+use Hubstaff\Auth;
+use Hubstaff\Decoder\DecodeDataInterface;
+use Hubstaff\Helper\ClientInterface;
+
+class AuthTest extends PHPUnit_Framework_TestCase
```
Comment (malukenho) [inline tests/AuthTest.php]: `final`

Comment (malukenho) [inline tests/AuthTest.php]: import `PHPUnit_Framework_TestCase`

File: tests/ActivitiesTest.php [hunk 83]
```diff
@@ -0,0 +1,115 @@
...
+            $startTime,
+            $stopTime,
+            'offset',
+            'options',
+            'url'
+        );
+
+        self::assertEquals(json_decode($expected, true), $actual);
+    }
+}
```
Comment (malukenho) [inline tests/ActivitiesTest.php]: EOL EOF

File: tests/ActivitiesTest.php [hunk 84]
```diff
@@ -0,0 +1,115 @@
...
+        $activities = new Activities($this->client, $this->decoder);
+
+        $actual = $activities->getActivities(
+            'auth_token',
+            'app_token',
+            $startTime,
+            $stopTime,
+            'offset',
+            'options',
+            'url'
```
Comment (malukenho) [inline tests/ActivitiesTest.php]: `,`

File: tests/ActivitiesTest.php [hunk 85]
```diff
@@ -0,0 +1,115 @@
+<?php
+
+namespace Hubstaff;
```
Comment (malukenho) [inline tests/ActivitiesTest.php]: test namespace

File: src/Helper/Curl.php [hunk 86]
```diff
@@ -0,0 +1,76 @@
...
+        return array($curl, $result);
+    }
+
+    /**
+     * @param $fields
+     * @param $parameters
+     *
+     * @return string
+     */
+    public function extractHeader($fields, $parameters)
```
Comment (malukenho) [inline src/Helper/Curl.php]: `array` ?

File: src/Helper/Curl.php [hunk 87]
```diff
@@ -0,0 +1,76 @@
...
+    {
+        $headers = array_filter(
+            $fields,
+            function ($key) use ($parameters) {
+                return isset($parameters[$key]) && $parameters[$key] === 'header';
+            },
+            ARRAY_FILTER_USE_KEY
+        );
+
+        // @todo to refactor later
```
Comment (malukenho) [inline src/Helper/Curl.php]: this need to be refactored

**Comment by malukenho** [comment]
@dexfs that seem pretty good, just need to split some concepts, fix some tests and take care about code style. Good JOB 🍻

File: tests/ActivitiesTest.php (hunk 85 above)
Comment (dexfs) [inline tests/ActivitiesTest.php]: Do you mean Hubstaff\Test  ?

**Comment by dexfs** [comment]
Thanks, there are a lot files that I didn't work and it was sent.   
I changed just Auth, Notes and Users, I'll work on others class and create a new PR

File: tests/UsersTest.php (hunk 47 above)
Comment (dexfs) [inline tests/UsersTest.php]: You're right. I saw after sent the message. 

But, why I need use a specific namespace for test ?

Comment (dexfs) [inline tests/UsersTest.php]: If I use HubstaffTest I will need import the tested classes, right ?

**Pull request closed without merging**